Fix cargo add behaving different when translating package name

dohse opened the pull request:
Fixes #13702
Fixes #10680

TODOs

- [x] ~Fuzzy match registry dependencies in `cargo remove` as well~
     `cargo remove` does not need fuzzy matching, because there is no unexpected behavior for the user
- [x] ~Don't duplicate name permutation generation~
     Unsure whether this is worth it
- [ ] Shall we reject cases where multiple different permutations match?
- [x] Add comments that explain the behavior

<!-- homu-ignore:start -->
<!--
Thanks for submitting a pull request 🎉! Here are some tips for you:

* If this is your first contribution, read "Cargo Contribution Guide" first:
  https://doc.crates.io/contrib/
* Run `cargo fmt --all` to format your code changes.
* Small commits and pull requests are always preferable and easy to review.
* If your idea is large and needs feedback from the community, read how:
  https://doc.crates.io/contrib/process/#working-on-large-features
* Cargo takes care of compatibility. Read our design principles:
  https://doc.crates.io/contrib/design.html
* When changing help text of cargo commands, follow the steps to generate docs:
  https://github.com/rust-lang/cargo/tree/master/src/doc#building-the-man-pages
* If your PR is not finished, set it as "draft" PR or add "WIP" in its title.
* It's ok to use the CI resources to test your PR, but please don't abuse them.

### What does this PR try to resolve?

Explain the motivation behind this change.
A clear overview along with an in-depth explanation are helpful.

You can use `Fixes #<issue number>` to associate this PR to an existing issue.

### How should we test and review this PR?

Demonstrate how you test this change and guide reviewers through your PR.
With a smooth review process, a pull request usually gets reviewed quicker.

If you don't know how to write and run your tests, please read the guide:
https://doc.crates.io/contrib/tests

### Additional information

Other information you want to mention in this PR, such as prior arts,
future extensions, an unresolved problem, or a TODO list.
-->
<!-- homu-ignore:end -->

epage commented on tests/testsuite/cargo_add/detect_workspace_inherit_fuzzy/mod.rs:
```diff
@@ -7,7 +7,6 @@ use cargo_test_support::str;
 use cargo_test_support::Project;
 
 #[cargo_test]
-#[ignore]
```
Please update your PR so the tests are never ignored and the first two commits pass on their own.  THat will make it so this commit will then change their output, showing the change in behavior.

dohse replied:
Ok, I updated the PR.

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -364,24 +364,28 @@ fn resolve_dependency(
-            // Assuming most existing keys are not relevant when the package changes
-            if selected_dep.optional.is_none() {
-                selected_dep.optional = old_dep.optional;
+    let get_dependency = |mut selected_dep: Dependency| {
```
> I created a fix for this issue. I did not follow the route of doing fuzzy lookups within cargo add, but just repeat the workspace existing dependency lookup and workspace dependency lookup when the latest dependency lookup returns a different name.

This approach is pretty complicated, requiring the code to duplicate its logic to cycle things back through.  At minimum, it seems like we could look in the registry from the beginning and then pass the literal name given and the registry normalized name around to do lookups.

Though that will also have problems.  This is using crates.io to normalize package names regardless of whether their source is crates.io or not.  We really should be normalizing as we go.  Unlike the registry, we don't need to try a bunch of combinations but we can extract the names from a table, normalize them to either `-` or `_`, normalize our value to `-` or `_` to match and see what the final result would be.

dohse replied:
Sorry, for the late reply.

> This approach is pretty complicated, requiring the code to duplicate its logic to cycle things back through.

I agree that this is not pretty. On the other hand it checks some boxes: It avoids duplicating logic in the code, It is pretty minimal, It avoids querying the registry if it is not needed, It applies the same fuzzy search logic as the registry.

> At minimum, it seems like we could look in the registry from the beginning and then pass the literal name given and the registry normalized name around to do lookups.

This has the problem that we always have to lookup the registry. This breaks a lot of tests by changing their outputs  by adding "Updating crates.io index" lines.

> This is using crates.io to normalize package names regardless of whether their source is crates.io or not.

Mhh, not sure whether I understand this properly. Sources of dependencies can be workspace dependencies, the members of the workspace and crates.io. As far as I understand only crates.io does fuzzy lookups. In which scenario crates.io be used to normalize package names that are not from crates.io?

> We really should be normalizing as we go.

I understand that you want to try several versions of the name - the original one, one that normalizes to underscores and one that normalizes to dashes, right? I can try that.

epage replied:
> Mhh, not sure whether I understand this properly. Sources of dependencies can be workspace dependencies, the members of the workspace and crates.io. As far as I understand only crates.io does fuzzy lookups. In which scenario crates.io be used to normalize package names that are not from crates.io?

git dependencies

> I understand that you want to try several versions of the name - the original one, one that normalizes to underscores and one that normalizes to dashes, right? I can try that.

That should be fine as that is all we are checking for the registry today, see #12083

dohse replied:
I updated this PR to check permutations

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -364,7 +364,24 @@ fn resolve_dependency(
 
-    let old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
+    let mut old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
+    if old_dep.is_none() && selected_dep.source().is_none() && selected_dep.rename().is_none() {
```
Why the `selected_dep` checks?

dohse replied:
`selected_dep.source().is_none()` to try the mutations only on when looking for registry dependencies and `selected_dep.rename().is_none()` to not try permutations when adding a renamed dependency.

epage replied:
Could you expand on the rename case?

dohse replied:
I changed the implementation - I believe this discussion is not relevant any more.

epage replied:
Please squash when you completely change directions in a PR.  It can be confusing to track what changed or didn't change between commits.

epage replied:
So while the code is changed, the concepts from this discussion are still there
```rust
    if let Some(rename) = dependency.rename() {
        return lookup(rename);
    }
```
```rust
            if !matches!(dep.source, Some(Source::Registry(_))) {
                continue;
            }
```

The motivation for these checks is not obvious.  Could you explain these?

Also, if there aren't tests that would break with their removal, could you add them?

dohse replied:
I added comments that explain the conditions.

When I remove the first condition these tests fail:

cargo_add::change_rename_target::case
cargo_add::invalid_key_rename_inherit_dependency::case
cargo_add::overwrite_rename_with_rename_noop::case

so that is covered.

That is not the case for the second condition, so I added a case to the test suite.

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -364,7 +364,24 @@ fn resolve_dependency(
     selected_dep = populate_dependency(selected_dep, arg);
 
-    let old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
+    let mut old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
+    if old_dep.is_none() && selected_dep.source().is_none() && selected_dep.rename().is_none() {
+        for name_permutation in [
+            selected_dep.name.replace('-', "_"),
+            selected_dep.name.replace('_', "-"),
+        ] {
```
Instead of looking up `old_dep` and then looking it up again, could we include the user-selected name in the list of permutations?

dohse replied:
I will try that, but I guess I have handle the rename case then separately.

epage replied:
Looks like you went ahead and did this in 543a138c285cec20f1454af02b4859cd90a03487 without any aspect of rename.  So is that not needed then?

dohse replied:
I am handling renames in [these lines](https://github.com/rust-lang/cargo/pull/13765/files#diff-7d08d5b8dd5fdeaca6bb19b814d06790d42f63f052b42a41f3eb5cf8ec89456cR463-R465) by not doing fuzzy lookups when the dependency is renamed

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -364,7 +364,24 @@ fn resolve_dependency(
-    let old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
+    let mut old_dep = get_existing_dependency(manifest, selected_dep.toml_key(), section)?;
+    if old_dep.is_none() && selected_dep.source().is_none() && selected_dep.rename().is_none() {
+        for name_permutation in [
+            selected_dep.name.replace('-', "_"),
+            selected_dep.name.replace('_', "-"),
+        ] {
```
Should we break this out into a function to be reused between this and `find_workspace_dep`?

dohse replied:
I did this and pushed a new version with that

torhovland commented:
This PR looks to be basically ready. @dohse, is there anything else you need to do before a final review? I see an unfinished TODO list in the description and one review comment that may not be resolved yet.

dohse commented:
Yes, this PR should be ready. I need feedback to my responses (@epage). I would take another look at my TODOs when the general direction of the fix is clear.

dohse commented:
I responded to your comments - I will also rebase this PR.

> In looking at the responses, its not clear to me what I needed responding to.

Sorry - this PR became confusing, because it contains two possible solutions. I was mostly referring to [this comment](https://github.com/rust-lang/cargo/pull/13765#discussion_r1589338337), that I broke out a function as suggested.

dohse commented:
I hope that I addressed the comment.

[I am also now only check permutations that are unique](https://github.com/rust-lang/cargo/pull/13765/files#diff-7d08d5b8dd5fdeaca6bb19b814d06790d42f63f052b42a41f3eb5cf8ec89456cR471)

epage commented on tests/testsuite/cargo_add/add_workspace_non_fuzzy/in/target/.rustc_info.json:
Looks like you accidentally pulled in your target directory

dohse replied:
Sorry - removed it

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -449,6 +452,45 @@ fn resolve_dependency(
+        };
+
+        if dependency.name != name_permutation {
+            // Ignore fuzzy matches with non-registry sources
```
This comment is a re-phrasing of the code and isn't describing the motivation behind it.

Based on the test, my guess is
```suggestion
            // Mirror the fuzzy matching policy of `PackageRegistry::query_vec`
```

dohse replied:
Sounds good.

I believe referencing `PackageRegistry::query()` is more clear, because `query_vec()` is only defined on the trait.

epage commented on src/cargo/ops/cargo_add/mod.rs:
```diff
@@ -449,6 +452,45 @@ fn resolve_dependency(
+    gctx: &GlobalContext,
+) -> CargoResult<Option<Dependency>> {
+    if let Some(rename) = dependency.rename() {
+        // If it has a rename only lookup that name - analogue to Dependency::toml_key()
```
This comment is a re-phrasing of the code and isn't describing the motivation behind it.

Oh, I think I see the key reason for this. Before we did lookups with `toml_key`.  This manually implements `toml_key` so that fuzzy lookups only happen on package names, not dependency names, to scope this to only happen when other normalization happens (`PackageRegistr::query_vec`).  Is that it?

dohse replied:
The intention is to mirror the behavior of `Dependency::toml_key()` to lookup _only_ on the rename if that is present. I agree with your description that fuzzy lookups should only happen on packages names.

I will write `Mirror behavior of to Dependency::toml_key()`

epage replied:
That describes the whole function in terms of checking `rename` and then `name`.

The specific nuance is why you are doing that, rather than using `toml_key`.

Something like
```
Manually implement `toml_key` to restrict fuzzy lookups to only package names to mirror `PackageRegistry::query_vec`
```

dohse replied:
I used the suggested comment.

dohse commented:
I changed the comments according to the discussion.

dohse commented:
I updated the comment.

I also removed the `.dedup()` of the combinations - it would not prevent all lookups of the unmodified original basename.

epage commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 9391bfb34d307ead02c00261de9f506feb45cb76 has been approved by `epage`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=epage 9391bfb34d307ead02c00261de9f506feb45cb76 -->
<!-- homu: {"type":"Approved","sha":"9391bfb34d307ead02c00261de9f506feb45cb76","approver":"epage","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 9391bfb34d307ead02c00261de9f506feb45cb76 with merge 1ed3596675d9338a0c215701051e64064b14b633...
<!-- homu: {"type":"BuildStarted","head_sha":"9391bfb34d307ead02c00261de9f506feb45cb76","merge_sha":"1ed3596675d9338a0c215701051e64064b14b633"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/10690747048/job/29636861307)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/cargo/actions/runs/10690747048/job/29636861307","builder_name":"checks-actions"} -->

epage commented:
@bors retry

bors commented:
:hourglass: Testing commit 9391bfb34d307ead02c00261de9f506feb45cb76 with merge 40b6638df6ee1360680a7577168eb1be4a167a83...
<!-- homu: {"type":"BuildStarted","head_sha":"9391bfb34d307ead02c00261de9f506feb45cb76","merge_sha":"40b6638df6ee1360680a7577168eb1be4a167a83"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/10692708913/job/29642512301)
Approved by: epage
Pushing 40b6638df6ee1360680a7577168eb1be4a167a83 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"epage","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/10692708913/job/29642512301"},"merge_sha":"40b6638df6ee1360680a7577168eb1be4a167a83"} -->
